mgmt: mcumgr: Add fs_mgmt log output workaround on file write

nordicjm opened the pull request:
If a file write was attempted with a file that does not exist, over
    the shell/UART when logging was enabled, it would output a fs_unlink
    error, this works around the issue by checking if the file exists and
    needs truncating before performing that action. It also imrproves
    flash endurance slightly by performing a truncate operation instead
    of a delete, but will fall back to a delete if the truncation
    operation fails. This issue can be also be mitigated by altering
    logging settings or adjusting the SMP thread priority.

Fixes #36882

Note that this only fixes the specific issue mentioned, there might be other potential errors emitted if filesystem errors are encountered, however the `mcumgr` client is actually to blame for this as it should be searching through the received data for the mcumgr start header, 0x06 0x09, which it does not do.

nordicjm committed:
mgmt: mcumgr: Add fs_mgmt log output workaround on file write

If a file write was attempted with a file that does not exist, over
the shell/UART when logging was enabled, it would output a fs_unlink
error, this works around the issue by checking if the file exists and
needs truncating before performing that action. It also imrproves
flash endurance slightly by performing a truncate operation instead
of a delete, but will fall back to a delete if the truncation
operation fails. This issue can be also be mitigated by altering
logging settings or adjusting the SMP thread priority.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

de-nordic reviewed (changes requested):
This does provide workaround that will work in 99% of the cases, but this is workaround.
I am not sure whether it should close the issue.

nordicjm commented:
> No it is not. The problem is that the client is in the middle of receiving frame that has been marked with start of header tag, but got additional characters inside. Because this is UART and CBOR a client basically sees stream of characters, following the start of header tag, and tries to process it as it is; unfortunately it gets something that is not part of packet in the middle so it fails.

From their log in the linked bug:
```
time="2021-07-27 10:56:02.502" level=debug msg="Rx serial:
00000000  1b 5b 31 3b 33 32 6d 75  61 72 74 3a 7e 24 20 1b  |.[1;32muart:~$ .|
00000010  5b 6d 1b 5b 38 44 1b 5b  4a 1b 5b 31 3b 33 32 6d  |[m.[8D.[J.[1;32m|
00000020  75 61 72 74 3a 7e 24 20  1b 5b 6d 06 09 41 42 55  |uart:~$ .[m..ABU|
00000030  44 41 41 41 4e 41 41 68  54 41 4c 39 69 63 6d 4d  |DAAANAAhTAL9icmM|
00000040  41 59 32 39 6d 5a 68 6b  48 67 76 39 6d 36 41 3d  |AY29mZhkHgv9m6A=|
00000050  3d                                                |=|
```
The 0x06 0x09 start header can be seen, and the full packet is there, I'm not sure where there isn't a 0x0a end character but other than that, the packet is not split up. The issue where UART data *does* interleave mcumgr commands from what I've seen is due to usage of `printk` - noted in https://github.com/zephyrproject-rtos/zephyr/issues/38947 the solution to that is to process printk messages in the logging system - PR at https://github.com/zephyrproject-rtos/zephyr/pull/50594 however this has had the side effect of causing many tests to fail to output any data, but in my testing, once that is enabled, shell/log/printk output does not interleave mcumgr packets

de-nordic commented:
So they did have specific case, but if you look at `uart_mcumgr_send_raw` you will notice that there are no exclusive locks on UART  and bytes are send one by one, additionally most of transitions on UART, by mcumgr, are done in chunks up to 4 bytes, by `mcumgr_serial_tx_small`, so there is plenty of time to jump in and make some mess.

de-nordic commented:
Don't get me wrong, this still fixes 99% of cases, probably, and I would accept it I am just asking whether we should mark it as workaround, with explanation, before closing the issue.
Disabling logging also does workaround for the problem, except in case when something else starts to write to uart.

nordicjm commented:
I remember having a look at trying to get locking working, but it was a free for all, shell has a mutex lock but logging doesn't, would probably make sense to implement a lock, although I think with some configurations it would end in a deadlock e.g. minimal footprint logging and mcumgr output, if a filesystem error is output during a UART mcumgr command runs. Alternatively, and this might not be a great idea, the smp thread priority could be changed to be negative/co-operative

nordicjm commented:
@nordic-krch could you weigh in on this, what would be the best path forward if there's multiple systems using the UART that need to lock it for periods of time so that they can output their data without interruption from other threads, and ideally where SMP operation would not cause any issues?

de-nordic commented:
> I remember having a look at trying to get locking working, but it was a free for all, shell has a mutex lock but logging doesn't, would probably make sense to implement a lock, although I think with some configurations it would end in a deadlock e.g. minimal footprint logging and mcumgr output, if a filesystem error is output during a UART mcumgr command runs. Alternatively, and this might not be a great idea, the smp thread priority could be changed to be negative/co-operative. Or perhaps use a UART output function that outputs the full buffer instead of a few bytes at a time, though this would have a RAM impact.

For now this is good as is, we just need to notify in commit that this is workaround, and there is alternative like disabling logging in FS.

nordicjm committed:
doc: release: 3.3: Add mcumgr fs_mgmt workaround issue note

Adds a note about a workaround for an issue with mcumgr whereby a
file write command is sent to the same UART used for logging.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

carlescufi commented:
@de-nordic please rebase

de-nordic commented:
@carlescufi Please review.
